src/main/java/org/testng/GuiceHelper.java

```diff
@@ -32,16 +39,57 @@ Injector getInjector(IClass iClass) {
 
     List<Module> moduleInstances =
         Lists.newArrayList(getModules(guice, parentInjector, iClass.getRealClass()));
+    List<Module> moduleLookup = Lists.newArrayList(moduleInstances);
+    Module parentModule = getParentModule(context);
+    if (parentModule != null) {
+      moduleInstances.add(parentModule);
+    }
 
-    // Reuse the previous injector, if any
-    Injector injector = context.getInjector(moduleInstances);
+    // Get an injector with the class's modules + any defined parent module installed
+    // Reuse the previous injector, if any, but don't create a child injector as JIT bindings can conflict
+    Injector injector = context.getInjector(moduleLookup);
     if (injector == null) {
-      injector = parentInjector.createChildInjector(moduleInstances);
+      injector = createInjector(context, moduleInstances);
+//      injector = parentInjector.createChildInjector(moduleInstances);
       context.addInjector(moduleInstances, injector);
     }
     return injector;
   }
 
+  private static Module getParentModule(ITestContext context) {
+    if (isStringEmpty(context.getSuite().getParentModule())) {
+      return null;
+    }
+    Class<?> parentModule = ClassHelper.forName(context.getSuite().getParentModule());
+    if (parentModule == null) {
+      throw new TestNGException(
+              "Cannot load parent Guice module class: " + context.getSuite().getParentModule());
+    }
+    if (!Module.class.isAssignableFrom(parentModule)) {
+      throw new TestNGException("Provided class is not a Guice module: " + parentModule.getName());
+    }
+    try {
+      Constructor<?> moduleConstructor = parentModule.getDeclaredConstructor(ITestContext.class);
+      return (Module)InstanceCreator.newInstance(moduleConstructor, context);
+    } catch (NoSuchMethodException e) {
+      return (Module)InstanceCreator.newInstance(parentModule);
+    }
+  }
+
+  public static Injector createInjector(ITestContext context, List<Module> moduleInstances) {
```

Comment:
This is a newly added method. Apart from `ClassImpl` there are no other invokers for this method. I see that `ClassImpl` is merely passing in an empty list. So can we not just remove off the `moduleInstances` parameter itself and just work with `List fullModules = Lists.newArrayList();`

Reply:
It's also used in this class at line 52.  I could add a convenience zero-arg overload for `ClassImpl` to call if you think that would be better?

Reply:
Oh yeah. I missed that. Then it's fine. Adding an overloaded version can be skipped